src/nvim/terminal.c

```diff
@@ -0,0 +1,986 @@
+      event_enable_deferred();
+    }
+
+    c = safe_vgetc();
+
+    if (process_deferred) {
+      event_disable_deferred();
+    }
+
+    switch (c) {
+      case Ctrl_BSL:
+        c = safe_vgetc();
+        if (c == Ctrl_N) {
+          goto end;
+        }
+        terminal_send_key(term, c);
+        break;
+
+      case K_LEFTMOUSE:
+      case K_LEFTDRAG:
+      case K_LEFTRELEASE:
+      case K_MIDDLEMOUSE:
+      case K_MIDDLEDRAG:
+      case K_MIDDLERELEASE:
+      case K_RIGHTMOUSE:
+      case K_RIGHTDRAG:
+      case K_RIGHTRELEASE:
+      case K_MOUSEDOWN:
+      case K_MOUSEUP:
+        if (send_mouse_event(term, c)) {
+          goto end;
+        }
+        break;
+
+      case K_EVENT:
+        event_process();
+        break;
+
+      default:
+        if (term->closed) {
+          close = true;
+          goto end;
+        }
+
+        terminal_send_key(term, c);
+    }
+
+    redraw();
+  }
+
+end:
+  term->focused = false;
+  State = save_state;
+  RedrawingDisabled = save_rd;
+  invalidate_terminal(term, cursor_line(term), cursor_line(term) + 1);
+  ui_unlock_cursor_state();
+  ui_cursor_on();
+  term->curwin = NULL;
+  mapped_ctrl_c = save_mapped_ctrl_c;
+  unshowmode(true);
+  redraw();
+  if (close) {
+    term->opts.close_cb(term->data);
+  }
+}
+
+void terminal_destroy(Terminal *term)
+{
+  term->buf->terminal = NULL;
+  pmap_del(ptr_t)(invalidated_terminals, term);
+  do_cmdline_cmd((uint8_t *)"bwipeout!");
+  for (size_t i = 0 ; i < term->sb_current; i++) {
+    free(term->sb_buffer[i]);
+  }
+  free(term->title);
+  free(term->sb_buffer);
+  vterm_free(term->vt);
+  free(term);
+}
+
+void terminal_set_data(Terminal *term, void *data)
+{
+  term->data = data;
+}
+
+void terminal_send(Terminal *term, char *data, size_t size)
+{
+  if (term->closed) {
+    return;
+  }
+  term->opts.write_cb(data, size, term->data);
+}
+
+void terminal_send_key(Terminal *term, int c)
+{
+  VTermModifier mod = VTERM_MOD_NONE;
+  VTermKey key = convert_key(c, &mod);
+
+  if (key) {
+    vterm_keyboard_key(term->vt, key, mod);
+  } else {
+    vterm_keyboard_unichar(term->vt, (uint32_t)c, mod);
+  }
+
+  size_t len = vterm_output_read(term->vt, term->textbuf,
+      sizeof(term->textbuf));
+  terminal_send(term, term->textbuf, (size_t)len);
+}
+
+void terminal_receive(Terminal *term, char *data, size_t len)
+{
+  if (!data) {
+    return;
+  }
+
+  vterm_input_write(term->vt, data, len);
+  vterm_screen_flush_damage(term->vts);
+}
+
+void terminal_get_line_attributes(Terminal *term, int line, int *term_attrs)
+{
+  int height, width;
+  vterm_get_size(term->vt, &height, &width);
+  int row = line - (int)term->sb_current - 1;
+
+  for (int col = 0; col < width; col++) {
+    VTermScreenCell cell;
+    if (row >= 0) {
+      vterm_screen_get_cell(term->vts, (VTermPos){.row = row, .col = col},
+          &cell);
+    } else {
+      // fetch the cell from the scrollback buffer
+      ScrollbackLine *sbrow = term->sb_buffer[-row - 1];
```

Comment:
For `line == 0` and `term->sb_current == 0`, the element `-row - 1 == 0` in an empty scrollback buffer is accessed.